hips/hip-0015.md

```diff
@@ -0,0 +1,126 @@
+      whitelisted: true
+```
+
+When a chart contains a dependency or subchart, the expection is that the metadata of the dependency/sub chart will contain the `helm.sh/images` annotation to identify the images required by the dependency chart. If that isn't true, the top level chart can provide the images entry for that chart and use the optional `dependency` field to identify the sub chart that the image is used in. For example, the Bitnami Kafka chart would identify the images used in `common` and `zookeeper` as part of the `kafka` chart metadat:
+
+```yaml
+annotations:
+  category: Infrastructure
+  images: |
+    - name: kafka
+      image: docker.io/bitnami/kafka:2.8.1-debian-10-r0
+    - name: kubectl
+      image: docker.io/bitnami/kubectl:1.19.5-debian-10-r3
+    - name: shell
+      image: docker.io/bitnami/bitnami-shell:10-debian-10-r199
+    - name: kafka-exporter
+      image: docker.io/bitnami/kafka-exporter:1.4.2-debian-10-r5
+    - name: jmx-exporter
+      image: docker.io/bitnami/bitnami/jmx-exporter:0.16.1-debian-10-r66
+    - name: zookeeper
+      image: docker.io/bitnami/zookeeper:3.7.0-debian-10-r157
+      dependency: zookeeper
+    - name: shell
+      image: docker.io/bitnami/bitnami-shell:10-debian-10-r202
+      dependency: zookeeper
+apiVersion: v2
+appVersion: 2.8.1
+dependencies:
+  - name: common
+    repository: https://charts.bitnami.com/bitnami
+    tags:
+      - bitnami-common
+    version: 1.x.x
+  - condition: zookeeper.enabled
+    name: zookeeper
+    repository: https://charts.bitnami.com/bitnami
+    version: 7.x.x
+description: Apache Kafka is a distributed streaming platform.
+engine: gotpl
+home: https://github.com/bitnami/charts/tree/master/bitnami/kafka
+icon: https://bitnami.com/assets/stacks/kafka/img/kafka-stack-220x234.png
+keywords:
+  - kafka
+  - zookeeper
+  - streaming
+  - producer
+  - consumer
+maintainers:
+  - email: containers@bitnami.com
+    name: Bitnami
+name: kafka
+sources:
+  - https://github.com/bitnami/bitnami-docker-kafka
+  - https://kafka.apache.org/
+version: 14.2.1
+```
+
+It's worth noting that the `bitnami-shell` image is used in both zookeeper and kafka, but with different tags.
+
+## Backwards compatibility
+
+Since there is no existing specification, the annotation will be optional to support backwards compatibility.
+
+## Security implications
+
+Have to think about it, but don't see any.
+
+## How to teach this
+
+The specification can be added to documentation around building helm charts, and exmaples of tools using this feature (e.g. ArtifactHub) would facilitate adoption/clarity on how its used.
+
+## Rejected Ideas
+
+### Helm Templating Discovery
+
+The methodology of dynamically discovering images used by a chart _can_ work in certain situations but also has limitations with images that are only used for certain use cases, as well as operator charts that don't expose the images needed as part of the rendered manifests.
```

Comment:
Note that images listed in CRs based on CRDs can't easily be discovered because we don't know how to navigate that schema. Determining the images is a bit of a guessing game. I am glad that Artifact Hub, at least, is trying to figure out some of those things.

Reply:
@joejulian are you suggesting putting a field for each image that specifies how an end user of the chart would override the image?

So for a chart that has
```yaml
image:
  registry: docker.io/bitnami/kafka
  tag: 2.8.1-debian-10-r0
...
```
The chart metadata would specify how to change the image and tag via:

```yaml
 annotations:
  images: |
    - name: kafka
      image: docker.io/bitnami/kafka:2.8.1-debian-10-r0
      registry: .image.registry 
      tag: .image.tag

Reply:
We do have a similar scenario but also to be able to pull images with digest, could we include digest as well.
```
 annotations:
  images: |
    - name: kafka
      image: docker.io/bitnami/kafka:2.8.1-debian-10-r0
      registry: .image.registry 
      tag: .image.tag
      digest: digest
```